### PR TITLE
test(tests): add tests for url.py

### PR DESCRIPTION
**Pull request by AlexMomot-717**

### CODE REVIEW

File: tests/fixtures/__init__.py
Comment (svntmr) [inline tests/fixtures/__init__.py:1]: no need for this package at the moment, we might not even use fixtures in this project

File: tests/paths.py
Comment (svntmr) [inline tests/paths.py:1]: can be removed if not used

File: tests/test_app.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+def test_validate():
+    # given
+    url = "https://www.weforum.org/my-forum/"
+
+    # when
+    errors = validate(url)
+
+    # then
+    expected_errors = {}
```
Comment (svntmr) [inline tests/test_app.py]: if expected result is not too big, you can inline it

File: tests/test_app.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    # when
+    normalized_url = normalize(url)
+
+    # then
+    expected_url = "http://www.example.com"
+    assert normalized_url == expected_url
+
+
+def test_check_url():
```
Comment (svntmr) [inline tests/test_app.py]: this test should be mocked as it'll fail without internet connection

File: tests/test_app.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        "продаже недвижимости Казахстана",
+        "description": "«Крыша» — тысячи свежих объявлений жилой "
+        "и коммерческой недвижимости в Казахстане. "
+        "Продажа, покупка и аренде квартир, домов, офисов, "
+        "участков по всему Казахстану.",
+    }
+    assert check_result == expected_result
+
+
+def test_check_url_not_exist():
```
Comment (svntmr) [inline tests/test_app.py]: same here

File: tests/test_app.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+    # when
+    check_result = check_url(url)
+
+    # then
+    expected_result = None
+    assert check_result == expected_result
+
+
+def test_check_url_unavailable():
```
Comment (svntmr) [inline tests/test_app.py]: same here

File: tests/test_app.py
Comment (svntmr) [inline tests/test_app.py:1]: this test module should be located at `tests/utils/test_url`

File: tests/test_app.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+
+def test_check_url():
+    # given
+    url = "https://krisha.kz/"
+
+    # when
+    check_result = check_url(url)
+
+    # then
+    expected_result = {
```
Comment (svntmr) [inline tests/test_app.py]: not necessary, but it'd be cool to see tests for `parse_page_data` function as well

**Review by svntmr — CHANGES_REQUESTED** [review]

File: pyproject.toml [hunk 6]
```diff
@@ -25,7 +25,10 @@ requests = "^2.31.0"
 beautifulsoup4 = "^4.12.3"
 
 [tool.poetry.group.dev.dependencies]
-pytest = "^7.4.4"
+pytest = "^8.0.1"
+coverage = "^7.4.2"
+pytest-cov = "^4.1.0"
+pytest-mock = "^3.12.0"
```
Comment (svntmr) [inline pyproject.toml]: not sure if it's used, can we remove it?

File: tests/utils/test_url.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+
+def test_validate_invalid():
+    # given
+    url = "htttps://www.weforum.org/my-forum/"
+
+    # when
+    errors = validate(url)
+
+    # then
+    expected_errors = {"name": "Некорректный URL"}
```
Comment (svntmr) [inline tests/utils/test_url.py]: [optional] if the expected result is not too big, you can inline it (applicable to other places)

File: tests/utils/test_url.py [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+    # when
+    normalized_url = normalize(url)
+
+    # then
+    expected_url = "http://www.example.com"
+    assert normalized_url == expected_url
+
+
+class TestCheckUrl(unittest.TestCase):
```
Comment (svntmr) [inline tests/utils/test_url.py]: it's not necessary to base your tests on `unittest.TestCase` class if you use patch or mock from `unittest`
tests in pytest are usually functions, not classes

File: tests/utils/test_url.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+    @patch("page_analyzer.utils.url.parse_page_data")
```
Comment (svntmr) [inline tests/utils/test_url.py]: wdyt about using [patch](https://docs.python.org/3/library/unittest.mock.html#quick-guide) as a context manager?
image

it looks better in the `given - when - then` tests setup IMO
and don't forget to check that mocks were called!

File: tests/utils/test_url.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+        mock_requests.get.return_value = request_response_mock
+        html_content = (
+            "<title>title</title>"
+            "<meta name='description' content='some description'></meta>"
+            "<h1>header</h1>"
+        )
+        request_response_mock.content = html_content
+        self.assertEqual(
+            "some description", parse_page_data(request_response_mock)["description"]
+        )
```
Comment (svntmr) [inline tests/utils/test_url.py]: wdyt about checking the whole returned object, not only one key? describing the full object will better describe code behaviour

**Review by svntmr — CHANGES_REQUESTED** [review]
closer!

File: tests/utils/test_url.py [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+    parsing_data = {
+        "response_code": "200",
+        "h1": "some header",
+        "title": "some title",
+        "description": "some description",
+    }
+
+    # when
+    with patch("page_analyzer.utils.url.parse_page_data") as mock_parse_page_data:
+        with patch("page_analyzer.utils.url.requests") as mock_requests:
```
Comment (svntmr) [inline tests/utils/test_url.py]: if you want to use two patches in one place (or two context managers in one place), you can do this:

```suggestion
    with patch("page_analyzer.utils.url.parse_page_data") as mock_parse_page_data, patch("page_analyzer.utils.url.requests") as mock_requests:
```

File: tests/utils/test_url.py [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+            request_response_mock = MagicMock()
+            request_response_mock.status_code = 200
+            mock_requests.get.return_value = request_response_mock
+            mock_parse_page_data.return_value = parsing_data
```
Comment (svntmr) [inline tests/utils/test_url.py]: usually only external stuff is mocked, but some heavy internal bits can be mocked as well
not sure if `parse_page_data` is a heavy part, wdyt? can we not mock it?

File: tests/utils/test_url.py [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+        "title": "some title",
+        "description": "some description",
+    }
+
+    # when
+    with patch("page_analyzer.utils.url.parse_page_data") as mock_parse_page_data:
+        with patch("page_analyzer.utils.url.requests") as mock_requests:
+            request_response_mock = MagicMock()
+            request_response_mock.status_code = 200
+            mock_requests.get.return_value = request_response_mock
```
Comment (svntmr) [inline tests/utils/test_url.py]: return values of patch objects can also be defined right in the patch context manager:

```
request_response = MagicMock()
request_response.status_code = 200
with patch("...", return_value=request_response) as ...:
    ...
```

File: tests/utils/test_url.py [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+            mock_parse_page_data.assert_called_once_with(request_response_mock)
+
+
+def test_check_url_exception():
+    # given
+    url = "http://www.fake.com"
+
+    # when
+    with patch("page_analyzer.utils.url.requests") as mock_requests:
+        mock_requests.get.side_effect = Exception
```
Comment (svntmr) [inline tests/utils/test_url.py]:
```suggestion
    with patch("page_analyzer.utils.url.requests", side_effect=Exception) as mock_requests:
```

File: tests/utils/test_url.py [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+
+        # then
+        with pytest.raises(Exception):
```
Comment (svntmr) [inline tests/utils/test_url.py]: if I remember correctly, `check_url` is designed to handle `requests.exceptions.RequestException` exceptions and return None, why do we raise plain `Exception` in this test?

File: tests/utils/test_url.py [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+
+def test_check_url_unavailable():
+    # given
+    url = "http://www.unavailable.com"
+
+    # when
+    with patch("page_analyzer.utils.url.requests") as mock_requests:
+        request_response_mock = MagicMock()
+        request_response_mock.status_code = 404
+        mock_requests.get.return_value = request_response_mock
```
Comment (svntmr) [inline tests/utils/test_url.py]:
```suggestion
    response = MagicMock()
    response.status_code = 404

    # when
    with patch("page_analyzer.utils.url.requests.get", 
 return_value=response) as mock_requests:
```

File: tests/utils/test_url.py [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+    request_response_mock.content = expected_html_content
+
+    # then
+    expected_result = {
+        "response_code": "200",
+        "h1": "header",
+        "title": "title",
+        "description": "description",
+    }
+    assert parse_page_data(request_response_mock) == expected_result
```
Comment (svntmr) [inline tests/utils/test_url.py]: let's also cover the case when `tags` is empty

**Review by svntmr — CHANGES_REQUESTED** [review]
really close!

File: tests/paths.py [hunk 18]
```diff
@@ -0,0 +1,7 @@
+from pathlib import Path
+
+TESTS_ROOT = Path(__file__).parent
+
+
+def resolve_path(file_location: str) -> Path:
```
Comment (svntmr) [inline tests/paths.py]: is it used now? if not, could we please remove it?

File: tests/utils/test_url.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+    normalized_url = normalize(url)
+
+    # then
+    assert normalized_url == "http://www.example.com"
+
+
+def test_check_url():
+    # given
+    url = "https://example.com"
+    parsing_data = {
```
Comment (svntmr) [inline tests/utils/test_url.py]: wdyt about moving it down to `# then` part?

File: tests/utils/test_url.py [hunk 20]
```diff
@@ -0,0 +1,145 @@
...
+        "<h1>some header</h1>"
+    )
+    request_response_mock = MagicMock()
+    request_response_mock.status_code = 200
+
+    # when
+    with patch(
+        "page_analyzer.utils.url.requests.get", return_value=request_response_mock
+    ) as mock_requests:
+        request_response_mock.content = expected_html_content
```
Comment (svntmr) [inline tests/utils/test_url.py]: let's move mock preparation before patch, like you do with the status code

File: tests/utils/test_url.py [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+
+        # then
+        expected_parsing_data = parsing_data
+        assert check_url(url) == expected_parsing_data
+        mock_requests.assert_called_once_with(url)
```
Comment (svntmr) [inline tests/utils/test_url.py]:
```suggestion
    with patch(
        "page_analyzer.utils.url.requests.get", return_value=request_response_mock
    ) as mock_requests:
        check_result = check_url(url)

        # then
        mock_requests.assert_called_once_with(url)
        assert check_result == {
            "response_code": "200",
            "h1": "some header",
            "title": "some title",
             "description": "some description",
        }
```

File: tests/utils/test_url.py [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+    url = "http://www.unavailable.com"
+
+    request_response_mock = MagicMock()
+    request_response_mock.status_code = 404
+
+    # when
+    with patch(
+        "page_analyzer.utils.url.requests.get", return_value=request_response_mock
+    ) as mock_requests:
+        mock_requests.get.return_value = request_response_mock
```
Comment (svntmr) [inline tests/utils/test_url.py]: is it necessary to assign a return value to the `get` property of the patch?

File: tests/utils/test_url.py [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+    request_response_mock.content = expected_html_content
+
+    # then
+    expected_result = {
+        "response_code": "200",
+        "h1": "header",
+        "title": "title",
+        "description": "description",
+    }
+    assert parse_page_data(request_response_mock) == expected_result
```
Comment (svntmr) [inline tests/utils/test_url.py]:
```suggestion
    html_content = (
        "<title>title</title>"
        "<meta name='description' content='description'></meta>"
        "<h1>header</h1>"
    )

    request_response_mock = MagicMock()
    request_response_mock.content = html_content

    # when
    page_data = parse_page_data(request_response_mock) 

    # then
    assert page_data == {
        "response_code": "200",
        "h1": "header",
        "title": "title",
        "description": "description",
    }
```

File: tests/utils/test_url.py [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+    request_response_mock.content = expected_html_content
+
+    # then
+    expected_result = {
+        "response_code": "200",
+        "h1": "header",
+        "title": "title",
+        "description": "",
+    }
+    assert parse_page_data(request_response_mock) == expected_result
```
Comment (svntmr) [inline tests/utils/test_url.py]:
```suggestion
    # given
    html_content = "<title>title</title><h1>header</h1>"

    request_response_mock = MagicMock()
    request_response_mock.content = html_content

    # when
    page_data = parse_page_data(request_response_mock)

    # then
    assert page_data == {
        "response_code": "200",
        "h1": "header",
        "title": "title",
        "description": "",
    }
```

**Review by svntmr — CHANGES_REQUESTED** [review]
final touches and we're ready to go!

File: tests/utils/test_url.py (hunk 22 above)
Comment (AlexMomot-717) [reply tests/utils/test_url.py:99]: forgot to remove

File: tests/utils/test_url.py [hunk 25]
```diff
@@ -0,0 +1,135 @@
...
+        check_result = check_url(url)
+
+        # then
+        mock_requests.assert_called_once_with(url)
+        assert check_result == {
+            "response_code": "200",
+            "h1": "some header",
+            "title": "some title",
+            "description": "some description",
+        }
```
Comment (svntmr) [inline tests/utils/test_url.py]: then part and patch asserts are usually done outside of patch context manager
```suggestion
    # then
    mock_requests.assert_called_once_with(url)
    assert check_result == {
        "response_code": "200",
        "h1": "some header",
        "title": "some title",
        "description": "some description",
    }
```

File: tests/utils/test_url.py [hunk 26]
```diff
@@ -0,0 +1,135 @@
...
+    url = "http://www.fake.com"
+
+    # when
+    with patch(
+        "page_analyzer.utils.url.requests.get",
+        side_effect=requests.exceptions.RequestException,
+    ) as mock_requests:
+        # then
+        assert check_url(url) is None
+        mock_requests.assert_called_once_with(url)
```
Comment (svntmr) [inline tests/utils/test_url.py]: then part and patch asserts are usually done outside of patch context manager
```suggestion
    # then
    assert check_url(url) is None
    mock_requests.assert_called_once_with(url)
```

File: tests/utils/test_url.py [hunk 27]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def test_check_url_unavailable():
+    # given
+    url = "http://www.unavailable.com"
+
+    request_response_mock = MagicMock()
+    request_response_mock.status_code = 404
```
Comment (svntmr) [inline tests/utils/test_url.py:94]: there should be a `patch` context manager, I guess?

**Review by svntmr — CHANGES_REQUESTED** [review]
a tiny bit left and we're good to go!

**Review by svntmr — APPROVED** [review]
good job!

**Pull request merged**